src/Operation/Watch.php

```diff
@@ -314,6 +326,7 @@ public function execute(Server $server)
         return new ChangeStream(
             $this->createChangeStreamIterator($server),
             fn ($resumeToken, $hasAdvanced): ChangeStreamIterator => $this->resume($resumeToken, $hasAdvanced),
+            $this->codec,
```

Comment:
`createChangeStreamIterator($server)` invokes `executeAggregate()`, which executes the Aggregate object created from the `$aggregateOptions` property. Notably, `$aggregateOptions` is never given the codec.

What's the reasoning for only passing it to the ChangeStream object? I noted that the ChangeStream applies the codec to its inner iterator (i.e. Aggregate cursor). Is it worth having the Aggregate operation apply the option itself since it already needs to support it? Or might that pose problems for accessing the resume token?

Reply:
If the `codec` option is passed to the `Aggregate` operation, the change stream iterator will receive the result of the codec operation. Once that happens, we can no longer make any assumptions about fields (which is why I purposefully used `id` as property for the `_id` field in the tests). The change stream iterator however needs to be able to extract the `_id` field to store it as a resume token. To allow for this, we don't pass the `codec` to the `Aggregate` operation, but set a `bson` type map on the resulting cursor if a codec is present. This allows us to extract the `_id` from the resulting document and then passing it on to the codec before returning it to the user.

I'll add that this might already be a problem if we pass a type map that results in an object that no longer allows accessing an `_id` property, but we haven't considered this so far.

---

While working on the `watch` operation, I was considering whether the codec given to the operation should be applied on the root level or whether we should wrap it in a special `ChangeStreamCodec` to ensure that we're handling the default document structure gracefully. I decided against this as users are able to provide their own pipeline, so we again can't make any assumptions about the documents returned in a change stream.

So, instead of not supporting codecs in `Watch` at all, I decided to implement them this way to allow users the flexibility of using them if they would like to. Given the complexity around change streams (especially with multiple events being returned that all have a different structure), I don't think users will be doing so very often, so I'm open to reconsider our approach regarding `Watch`.

Reply:
> I'll add that this might already be a problem if we pass a type map that results in an object that no longer allows accessing an _id property, but we haven't considered this so far.

Good point. Happy to wait until that bug report comes in, if it ever does.

----

Quoting the [Change Stream spec](https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#helper-method):

> The server will produce an error if `_id` is projected out, but a user should otherwise be able to modify the shape of the change stream event as desired.

I don't think we need to worry about the `_id` not existing; however, it becoming inaccessible due to a `typeMap` is a legitimate (albeit unlikely) concern.

Reply:
Agree about not worrying about this until it comes in. Since Codecs might potentially make type maps obsolete, there's a very good chance that this won't happen at all.